pkg/operator/observe_config.go

```diff
@@ -0,0 +1,147 @@
+package operator
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/golang/glog"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/diff"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/client-go/util/workqueue"
+
+	operatorconfigclientv1alpha1 "github.com/openshift/cluster-kube-controller-manager-operator/pkg/generated/clientset/versioned/typed/kubecontrollermanager/v1alpha1"
+	operatorconfiginformerv1alpha1 "github.com/openshift/cluster-kube-controller-manager-operator/pkg/generated/informers/externalversions/kubecontrollermanager/v1alpha1"
+)
+
+type observeConfigFunc func(kubernetes.Interface, *rest.Config, map[string]interface{}) (map[string]interface{}, error)
+
+type ConfigObserver struct {
+	operatorConfigClient operatorconfigclientv1alpha1.KubecontrollermanagerV1alpha1Interface
+
+	kubeClient   kubernetes.Interface
+	clientConfig *rest.Config
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+
+	rateLimiter flowcontrol.RateLimiter
+	observers   []observeConfigFunc
+}
+
+func NewConfigObserver(
+	operatorConfigInformer operatorconfiginformerv1alpha1.KubeControllerManagerOperatorConfigInformer,
+	kubeInformersForOpenShiftKubeControllerManagerNamespace informers.SharedInformerFactory,
+	operatorConfigClient operatorconfigclientv1alpha1.KubecontrollermanagerV1alpha1Interface,
+	kubeClient kubernetes.Interface,
+	clientConfig *rest.Config,
+) *ConfigObserver {
+	c := &ConfigObserver{
+		operatorConfigClient: operatorConfigClient,
+		kubeClient:           kubeClient,
+		clientConfig:         clientConfig,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConfigObserver"),
+
+		rateLimiter: flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
+		observers:   []observeConfigFunc{},
```

Comment:
Will the observeConfigFunc implementations be in a follow-up?

Reply:
> Will the observeConfigFunc implementations be in a follow-up?

I think that will be your task and @juanvallejo can review.